### DISCUSSION TITLE
Bed probe positions.

### DISCUSSION THREAD

**Discussion by GilesBathgate** [vintagepc/MK404#341] [General]
Hi, I don't understand how these positions are derived.

https://github.com/vintagepc/MK404/blob/029cc59a8de950150892847766d46e4fb6ebb9ba/parts/components/PINDA.cpp#L186-L189

We have the definitions here:

https://github.com/prusa3d/Prusa-Firmware/blob/MK2/Firmware/mesh_bed_calibration.cpp#L53

```c++
const float bed_ref_points[] PROGMEM = {
    13.f  - BED_ZERO_REF_X,   6.4f - BED_ZERO_REF_Y,
    115.f - BED_ZERO_REF_X,   6.4f - BED_ZERO_REF_Y,
 ```
 
 but `BED_ZERO_REF_X` is `#define BED_ZERO_REF_X (- 22.f + X_PROBE_OFFSET_FROM_EXTRUDER)` 
 and `BED_ZERO_REF_Y` is `#define BED_ZERO_REF_Y (- 0.6f + Y_PROBE_OFFSET_FROM_EXTRUDER)`
 
 I get that you set the pinda offset from extruder as 0,0, so ignore those, but where does the magic `3` come from? surely it would be:
 
 ```c++
 static float _fMK2Cal[18] = { 
 	 13.f -  -22.f,	  6.4f - -0.6f, 
 	 13.f - -22.f, 	104.4f - 0.6f, 
 	 //...
 ```
 ?! :man_shrugging:

**Comment by vintagepc** [comment]
That is the Y offset. In the firmware there's some trickery that makes the bed front (prime line) be at Y = -3, out of the normal printable region - the bed is actually a few mm longer than the printable area. 

This offset is to correct between the simulated motor position, which goes from 0 to 213 to place the probe points in the correct region from 0-210. In other words, the values in the firmware are relative to bed zero, whereas the simulator values are relative to the motor travel "zero".

**Reply by GilesBathgate** [reply]
Ok that makes sense I think. Although the magic numbers in the code could do with a comment ;)

In the firmware the Y position can go to a negative value.  
https://github.com/prusa3d/Prusa-Firmware/blob/MK2/Firmware/variants/1_75mm_MK2-RAMBo13a-E3Dv6full.h#L57
```c++
#define Y_MIN_POS -2.2
```
I am basically testing [this change](https://github.com/GilesBathgate/Prusa-Firmware/commit/b738fe9fb06f5911877a48fdab0563f0236983ce) to the MK3 firmware to get it running on my MK2 which I was discussing with @leptun here:
https://github.com/prusa3d/Prusa-Firmware/commit/8ef87d76ef46379586f8248b1e61b0232f02d346

**Reply by vintagepc** [reply]
aha :) 

If I remember correctly @3d-gussner  was experimenting with something similar and also had issues with bed calibration failing; I can't recall the details but there is some voodoo somewhere we don't quite understand because even though I thought the math checked out it would fail the skew correction... I want to say the same problem happens on real hardware but I'm not 100% sure because this was quite a few months ago...

**Reply by GilesBathgate** [reply]
@vintagepc That's the thing, the problem doesn't happen on real hardware with my patch. Having just implemented it, and tested it on my MK2 I wanted to find some kind of unit test framework ;)

**Reply by vintagepc** [reply]
ah, interesting... IIRC there is a way in the FW to get more debug output from the calibration process to figure out exactly why it's unhappy. We're probably failing some of the positioning geometry with our best-guess point locations; they've been reverse engineered rather than produced based on any sort of scientific measurements.  I will readily admit the sim is not perfect and probably has some bugs here and there that we have not found yet. 

The other thing I can think of is the probe point shape; the calibration is somewhat sensitive to that and in the past if the "point" it finds is too steep or not steep enough it will also fail.

**Reply by GilesBathgate** [reply]
@vintagepc Ah solved my issue.... In your sim you have 
https://github.com/vintagepc/MK404/blob/ad578c36c8ec6d0fc226ac973c98383b59a0432d/parts/components/A4982.cpp#L77

This prevents the Y position from ever going negative. If I comment out the line my patch works.

**Reply by vintagepc** [reply]
Simulated motor positions should never go negative. This would suggest the axis length might be incorrect or that "home" is not at zero then and the printer expects to be able to travel slightly more once the endstop is triggered.

**Reply by vintagepc** [reply]
Any idea how many steps it tries to overtravel?

**Reply by GilesBathgate** [reply]
Sorry, the simulated motor positions don't go negative. Only the firmware Y.
![image](https://user-images.githubusercontent.com/444143/143659530-ea53d0c9-ac11-42ba-b028-6be5f9294d15.png)

Previously X and Y were displaying 10

**Reply by GilesBathgate** [reply]
Hmm, if I add a check for the Axis and ensure z cannot go negative I have the same issue. So not the correct fix :(

**Reply by vintagepc** [reply]
What happens if you change the tripping point for the MIN_OUT IRQ from <0 to something like 0.5/1mm (50/100)?

That way the endstop triggers before we run off the end of the motor travel but leaves some "squishy' homing room.

**Comment by GilesBathgate** [comment]
@vintagepc Well, I think I've found a bug. (maybe an old MK2 firmware bug, or simulation bug. )
I use a stock MK2S 3.2.3 firmware image (from prusa download page https://help.prusa3d.com/en/downloads/)

`$ ./MK404  -f ~/1_75mm_MK2-RAMBo13a-E3Dv6full.hex Prusa_MK2_mR13`

If I run selftest immediately after starting the printer up, no problems, but if I run auto home first, and then selftest:

![image](https://user-images.githubusercontent.com/444143/143674156-d7b78426-1fba-4aba-b1a7-15a5a3b48535.png)

**Comment by vintagepc** [comment]
Hmm... that almost sounds like the firmware is not moving away from the axis end after a home and allowing the virtual endstop to "untrigger". Then when it attempts selftest the endstop is already set and it doesn't like that.

**Comment by GilesBathgate** [comment]
@vintagepc The bad news is that this is not a firmware bug. I've just flashed stock MK2 to my physical MK2, and it seems to work fine.

What I can say though is that after an autohome, if I click Settings -> Move Y, the reading is `-001.1`. If I then do the same on the sim, I get `-001.9`

**Comment by vintagepc** [comment]
iiiiiiiiiiiiiiinteresting. I'll see if I can find some time to look into what's going on today... unless you beat me to it :rofl: 

might be interesting to do a VCD dump of the step count and step pins to see if we're somehow continuing to step into a region we shouldn't and hitting undefined behaviour

**Reply by GilesBathgate** [reply]
>Do a VCD dump of the step count and step pins 

I tried `./MK404 Prusa_MK2_mR13 -t "A4982_Y_<A4982.step"` I am not sure where I am expecting the output to appear.

**Reply by vintagepc** [reply]
It should create a .vcd file in the same directory that can be opened/inspected with e.g. GTKWave.

**Reply by GilesBathgate** [reply]
```
$ strace ./MK404 Prusa_MK2_mR13 -t "A4982_Y_<A4982.step" 2>&1 | grep "\.vcd" & sleep 10 && ls *.vcd
[2] 27797
Initialized VCD file Prusa_MK2_13_VCD.vcd
ls: cannot access '*.vcd': No such file or directory
```

**Reply by vintagepc** [reply]
If memory serves, the file is written in chunks with some internal buffering so if there's not a lot of data, it may not appear until the sim is closed. 

In any case, start with the -trace option, trigger the failure, quit the sim and then it should appear. Sorry for the confusion - some of this stuff isn't well documented here because it's not our feature we're leveraging.

**Reply by GilesBathgate** [reply]
nope, no VCD file, I did an autohome, selftest, then quit. No file. :(
I will wait for you to try and reproduce. Thanks.

**Reply by vintagepc** [reply]
I was able to reproduce the fault. 

What fixed it for me was changing this in `A4982.cpp::CheckEndstops()`:
`if (m_iCurStep<0)` to `if (m_iCurStep<=0)`

Does that fix it for you?

**Reply by vintagepc** [reply]
... also, I'm an idiot. I forgot to tell you that because VCDs are very spammy and can balloon quickly it doesn't actually record anything until toggled on with the `+` key (and `-` will stop logging)
![image](https://user-images.githubusercontent.com/53943260/143688977-b18c848f-4eb2-4525-a297-a29e68364d6e.png)

**Reply by GilesBathgate** [reply]
> I was able to reproduce the fault.
> 
> What fixed it for me was changing this in `A4982.cpp::CheckEndstops()`: `if (m_iCurStep<0)` to `if (m_iCurStep<=0)`
> 
> Does that fix it for you?

That fixes the selftest after autohome for stock MK2 yeah, I've yet to try my patched firmware though. Should there be an equivalent fix for TMC2130?

**Reply by vintagepc** [reply]
I haven't observed the same issue with the TMC2130, probably because the endstops behave a little differently - the DIAG lines will go back to low after a certain amount of time automatically if you don't continue running into the end of the travel and skipping steps.

**Reply by GilesBathgate** [reply]
@vintagepc The good news is, that with my patched firmware if I do autohome/selftest it now works. The bad news is that if I just start the printer and do selftest I still have an issue, and the Ugly news is that this is the exact opposite behaviour with my physical MK2, that is to say if I do a autohome/selftest it fails, so this is good, your sim helped me find a bug :)

I'll try and do the vcd thing, but I am not really sure what to do. Could you take a look at my hex 
[FW-Build4481-1_75mm_MK2-RAMBo13a-E3Dv6full-EN_ONLY.zip](https://github.com/vintagepc/MK404/files/7612210/FW-Build4481-1_75mm_MK2-RAMBo13a-E3Dv6full-EN_ONLY.zip) 
(or build it from: https://github.com/GilesBathgate/Prusa-Firmware/tree/MK2-3.10.x)

```
./MK404 Prusa_MK2_mR13 -f ~/FW-Build4481-1_75mm_MK2-RAMBo13a-E3Dv6full-EN_ONLY.hex 
```

**Reply by GilesBathgate** [reply]
@vintagepc So the fix for my patched firmware seems to be: https://github.com/GilesBathgate/Prusa-Firmware/commit/81ba67bd08f42e37f2b333fa2eecf45be388ea2f I still have no idea why, but it completely ignores the XY move unless it's done separately.

**Reply by vintagepc** [reply]
Interesting. Alas I'm not an expert in that area of the FW so no idea why here either...

**Reply by GilesBathgate** [reply]
Maybe @leptun has an idea, (if you are not too busy)

**Reply by leptun** [reply]
Sorry, no idea. The offsets and calibration points have always confused me and honestly made little sense. I'm too afraid to touch that code :)

**Reply by GilesBathgate** [reply]
Well thanks for taking a look. I guess its either a bug in the sim, or a problem with my understanding. Maybe I will get to the bottom of it someday ;)

**Reply by GilesBathgate** [reply]
@leptun So it seems to be that when the endstops are enabled the XY move doesn't happen in the sim, but it does happen on real hardware. This change fixed it: https://github.com/GilesBathgate/Prusa-Firmware/commit/6c2330501e74eea87ae8de47f53dc333c1041a3b
